harness/determined/searcher/core_search_runner.py

```diff
@@ -0,0 +1,121 @@
+import logging
+import os
+import pickle
+import time
+from typing import Any, Dict, List, Optional, Tuple
+
+import determined as det
+from determined.common.api.bindings import (
+    determinedexperimentv1State,
+    get_GetExperiment,
+    post_PauseExperiment,
+)
+from determined.experimental import client
+from determined.searcher.search_method import Operation, SearchMethod
+from determined.searcher.search_runner import SearchRunner
+
+
+class CoreSearchRunner(SearchRunner):
+    def __init__(self, search_method: SearchMethod, context: det.core.Context) -> None:
+        super().__init__(search_method)
+        self.context = context
+        self.info = det.get_cluster_info()
+
+        assert self.info is not None, "CoreSearchRunner only runs on-cluster"
```

Comment:
You are saving an `Optional[ClusterInfo]` object as `self.info`, which is why you have to have other assertions in other places.

Instead, assert not None before saving to `self.info`, and mypy will know self.info is of type `ClusterInfo`:

```python
info = det.get_cluster_info()
assert info is not None, "..."
self.info = info
```